feat: Better error reporting when features are disabled

thomaseizinger opened the pull request:
## Description

<!-- Please write a summary of your changes and why you made them.-->

In case support for e.g. RSA keys is disabled at compile-time, we will now print a better error message. For example: 

> Failed to dial Some(PeerId("QmcZf59bWwK5XFi76CZX8cbJ4BhTzzA3gU1ZjYZcYW3dwt")): Failed to negotiate transport protocol(s): [(/dnsaddr/bootstrap.libp2p.io/p2p/QmcZf59bWwK5XFi76CZX8cbJ4BhTzzA3gU1ZjYZcYW3dwt): : Handshake failed: Handshake failed: Invalid public key: Key decoding error: RSA keys are unsupported)]

Fixes #2971.

## Links to any relevant issues

<!-- Reference any related issues.-->

Depends-On: https://github.com/libp2p/rust-libp2p/pull/3081

## Open Questions

<!-- Unresolved questions, if any. -->

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

thomaseizinger committed:
Return `snow::Error` as source

This TODO can be resolved now.

thomaseizinger committed:
Use `thiserror` macro to generate error

thomaseizinger committed:
Move `NoiseError` to `lib.rs`

Having an entire module for one struct doesn't quite cut it.

thomaseizinger committed:
Be more specific about error case for bad signatures

thomaseizinger committed:
Be more specific about error case for unexpected key in authentication

thomaseizinger committed:
Use `?` for better indentation

thomaseizinger committed:
Even more use of `?`

thomaseizinger committed:
Use combinators on ALL the things

thomaseizinger committed:
Be more specific about error case when validating key length

thomaseizinger committed:
Include source of decoding error

This should make for better error messages when support for certain
keys is disabled.

Fixes #2971.

thomaseizinger committed:
Add changelog entry

thomaseizinger committed:
Slightly revise coding style of `ipfs-kad` example

thomaseizinger committed:
Be more specific about what is not supported

thomaseizinger committed:
Don't include sources in Display impl

This leads to double printing of errors when iterating the source
chain.

thomaseizinger committed:
Print source chain on `DialError`

thomaseizinger committed:
Format code

thomaseizinger committed:
Fix bad parenthesis

thomaseizinger committed:
Fix clippy error

mxinden commented on core/src/identity.rs:
```diff
@@ -323,7 +323,7 @@ impl TryFrom<keys_proto::PublicKey> for PublicKey {
             keys_proto::KeyType::Rsa => {
                 log::debug!("support for RSA was disabled at compile-time");
-                Err(DecodingError::new("Unsupported"))
+                Err(DecodingError::new("RSA keys are unsupported"))
```
``` diff
-                Err(DecodingError::new("RSA keys are unsupported"))
+                Err(DecodingError::new("RSA keys are unsupported. Did you enable the `rsa` feature?"))
````

How about mentioning the feature flag here and below?

mxinden replied:
Or maybe "Consider enabling the `rsa` feature.".

thomaseizinger replied:
Thanks, that gave me an idea for a better wording!

thomaseizinger replied:
See [`c81f730` (#2972)](https://github.com/libp2p/rust-libp2p/pull/2972/commits/c81f730aaceef796cae4ebe09155cb318f4095bb).

mxinden commented on swarm/src/lib.rs:
```diff
@@ -2502,4 +2526,23 @@ mod tests {
+
+        let string = format!("{error}");
+
+        // Unfortunately, we have some "empty" errors that lead to multiple colons without text but that is the best we can do.
```
What is an empty error?

thomaseizinger replied:
An error that does not write anything on `Display::fmt`.

mxinden reviewed (commented):
As always, great work. Thank you for tackling this so thoroughly. Thanks for breaking it up into smaller meaningful commits.

thomaseizinger committed:
Introduce dedicated constructors for `DecodingError`

mxinden committed:
Merge branch 'master' into 2971-better-error-message

mxinden committed:
Merge branch 'master' into 2971-better-error-message

mxinden committed:
Merge branch 'master' into 2971-better-error-message

thomaseizinger committed:
Merge branch 'master' into 2971-better-error-message

thomaseizinger committed:
Merge branch 'master' into 2971-better-error-message

thomaseizinger committed:
Merge branch 'master' into 2971-better-error-message

thomaseizinger committed:
Bump version accordingly

mxinden reviewed (approved):
`libp2p-noise` version bump looks good to me.

Either this pull request or some other pull request needs to bump the `libp2p-core` version. Otherwise this change will not be released within the next release round.

Also I don't we should merge here before https://github.com/libp2p/test-plans/issues/59.

Otherwise this looks good to me. Thanks. And again, I wished I merged this before the last release.

thomaseizinger committed:
Merge branch 'master' into 2971-better-error-message

thomaseizinger committed:
Increase `libp2p-core` version to 0.37.1

mxinden committed:
Merge branch 'master' into 2971-better-error-message

thomaseizinger committed:
Merge branch 'master' into 2971-better-error-message

thomaseizinger committed:
Merge branch 'master' into 2971-better-error-message

thomaseizinger committed:
Fix bad boolean logic

We only need this if one of these is not enabled, i.e. they are
`not(all)` enabled.

thomaseizinger committed:
Merge branch 'master' into 2971-better-error-message

thomaseizinger committed:
Merge branch 'master' into 2971-better-error-message

thomaseizinger committed:
Remove unnecessary feature activation

This is now released and available on crates.io.

thomaseizinger committed:
Revert "Remove unnecessary feature activation"

This reverts commit cedd5a164622f0850688da2aff0bb4222e02629e.

thomaseizinger committed:
Fix semver problem

melekes reviewed (approved):
👍

thomaseizinger committed:
MOAR CONDITIONALS

thomaseizinger committed:
Fmt
